domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt

```diff
@@ -0,0 +1,291 @@
+package org.oppia.android.domain.exploration.lightweightcheckpointing
+
+import kotlinx.coroutines.Deferred
+import org.oppia.android.app.model.ExplorationCheckpoint
+import org.oppia.android.app.model.ExplorationCheckpointDatabase
+import org.oppia.android.app.model.ExplorationCheckpointDetails
+import org.oppia.android.app.model.ProfileId
+import org.oppia.android.data.persistence.PersistentCacheStore
+import org.oppia.android.domain.oppialogger.OppiaLogger
+import org.oppia.android.util.data.AsyncResult
+import org.oppia.android.util.data.DataProvider
+import org.oppia.android.util.data.DataProviders
+import org.oppia.android.util.data.DataProviders.Companion.transformAsync
+import javax.inject.Inject
+import javax.inject.Singleton
+
+private const val CACHE_NAME = "exploration_checkpoint_database"
+private const val RETRIEVE_EXPLORATION_CHECKPOINT_DATA_PROVIDER_ID =
+  "retrieve_exploration_checkpoint_provider_id"
+private const val RETRIEVE_OLDEST_CHECKPOINT_DETAILS_DATA_PROVIDER_ID =
+  "retrieve_oldest_checkpoint_details_provider_id"
+private const val RECORD_EXPLORATION_CHECKPOINT_DATA_PROVIDER_ID =
+  "record_exploration_checkpoint_provider_id"
+private const val DELETE_EXPLORATION_CHECKPOINT_DATA_PROVIDER_ID =
+  "delete_exploration_checkpoint_provider_id"
+
+/**
+ * Controller for saving, retrieving, updating, and deleting exploration checkpoints.
+ */
+@Singleton
+class ExplorationCheckpointController @Inject constructor(
+  private val cacheStoreFactory: PersistentCacheStore.Factory,
+  private val dataProviders: DataProviders,
+  private val oppiaLogger: OppiaLogger,
+  @ExplorationStorageDatabaseSize private val explorationCheckpointDatabaseSizeLimit: Int
+) {
+
+  /** Different stages in which the exploration checkpoint database can exist. */
+  enum class ExplorationCheckpointDatabaseState {
```

Comment:
@BenHenning @rt4914 The enum class here and below and replicated should we use a single enum class here or using two different classes is fine? I prefer we should use a single enum class instead of duplicating them for the particular scenario.

Reply:
@aggarwalpulkit596 Yes we can wrap this into single enum.